### PR TITLE
Json encoders/decoders

### PR DESCRIPTION
**Pull request by mattjbray**

### CODE REVIEW

File: .github/workflows/main.yml [hunk 1]
```diff
@@ -12,7 +12,7 @@ jobs:
     - uses: actions/checkout@master
     - uses: avsm/setup-ocaml@v2
```
Comment (c-cube) [inline .github/workflows/main.yml]:
```suggestion
    - uses: ocaml/setup-ocaml@v2
```

File: src/document.ml [hunk 2]
```diff
@@ -357,8 +357,8 @@ module Graph = struct
 
   type t = event list
 
-  let n ?lbl name = Node {name; label=CCOpt.get_or ~default:name lbl}
-  let e ?lbl src target = Edge {src;target; label=CCOpt.get_or ~default:"" lbl}
+  let n ?lbl name = Node {name; label=CCOption.get_or ~default:name lbl}
+  let e ?lbl src target = Edge {src;target; label=CCOption.get_or ~default:"" lbl}
```
Comment (c-cube) [inline src/document.ml:361]: we could use `Option` now, with recent versions. Oh well.

File: src/json/imandra_document_json.ml [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+    | See_file s -> obj_ty "file" [ ("value", string s) ]
+    | See_doc s -> obj_ty "doc" [ ("value", string s) ]
+
+  let rec t (doc : Document.t) =
+    obj
+      (("attributes", attributes (Document.attrs doc))
+      :: view (Document.view doc))
+
+  and view (view : Document.view) =
+    let with_ty ty kvs = ("ty", string ty) :: kvs in
```
Comment (c-cube) [inline src/json/imandra_document_json.ml:36]: :heart_eyes:

File: src/json/imandra_document_json.ml [hunk 4]
```diff
@@ -0,0 +1,243 @@
...
+    | "return_value" ->
+        field "doc" t >>= fun doc -> succeed (Document.OT_return_value doc)
+    | "inline" -> succeed Document.OT_inline
+    | "custom" ->
+        field "custom" string >>= fun custom ->
+        field "doc" t >>= fun doc -> succeed (Document.OT_custom (custom, doc))
+    | "canonical" ->
+        field "canonical" string >>= fun canonical ->
+        succeed (Document.OT_canonical canonical)
+    | _ -> fail "unknown ocamldoc_tag type"
```
Comment (c-cube) [inline src/json/imandra_document_json.ml:166]:
```suggestion
    | _tag -> fail (Printf.sprintf "unknown ocamldoc_tag type: %S" _tag)
```

File: src/json/imandra_document_json.ml [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        field "html" string >>= fun html ->
+        succeed (Document.html (Document.Unsafe_.html_of_string html))
+    | "record" ->
+        let item =
+          field "key" string >>= fun key ->
+          field "value" t >>= fun value -> succeed (key, value)
+        in
+        field "record" (list item) >>= fun items ->
+        succeed (Document.record ~a items)
+    | _ -> fail "unknown document type"
```
Comment (c-cube) [inline src/json/imandra_document_json.ml:239]:
```suggestion
    | _tag -> fail (Printf.sprintf "unknown document type: %S" _tag)
```

**Review by c-cube — COMMENTED** [review]
Just a few small changes, but will merge after them.

File: src/json/imandra_document_json.ml (hunk 4 above)
Comment (mattjbray) [reply src/json/imandra_document_json.ml:166]: Decoders already adds the context to the error so no need.

**Pull request merged**